base/parallel.cc

```diff
@@ -0,0 +1,122 @@
+/** parallel.cc                                                    -*- C++ -*-
+    Jeremy Barnes, 5 February 2016
+    Copyright (c) 2016 Datacratic Inc.  All rights reserved.
+
+    This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.
+*/
+
+#include "parallel.h"
+#include "mldb/compiler/compiler.h"
+#include "mldb/base/exc_assert.h"
+#include "thread_pool.h"
+#include <atomic>
+#include <mutex>
+
+namespace Datacratic {
+
+void parallelMap(size_t first, size_t last,
+                 const std::function<void (size_t)> & doWork,
+                 int occupancyLimit)
+{
+    ExcAssertGreaterEqual(last, first);
+    ExcAssertLess((last - first), 1ULL << 31);
+
+    std::atomic<int> has_exc(0);
+    std::atomic<size_t> index(first);
+    std::mutex exc_mutex;
+    std::exception_ptr exc;
+
+    // This creates a thread pool that runs jobs on the default thread pool
+    ThreadPool tp;
+
+    if (occupancyLimit == -1)
+        occupancyLimit = numCpus();
+    if (occupancyLimit > (last - first))
+        occupancyLimit = (last - first);
+
+    auto worker = [&] ()
+        {
+            while (!has_exc.load(std::memory_order_relaxed)) {
+                size_t myindex = index.fetch_add(1);
+                if (myindex >= last)
+                    return;
+                try {
+                    doWork(myindex);
+                } JML_CATCH_ALL {
+                    if (has_exc.fetch_add(1) == 0) {
+                        ExcAssert(!exc);
+                        exc = std::current_exception();
```

Comment:
The first one is captured, and the rest are caught and silently ignored.  I guess it would be possible to make it capture all exceptions, except that the number caught would vary because no new jobs are started once the first has been detected.